File: Java/10_binary_tree/_226_Invert_Binary_Tree.java
import java.util.List;

import static java.util.Arrays.asList;

public class _226_Invert_Binary_Tree {
    public static void main(String[] args) {

        // Test 1
        TreeNode root = buildTree(
                asList(4,
                        asList(2,
                                List.of(1),
                                List.of(3)
                        ),
                        asList(7,
                                List.of(6),
                                List.of(9)
                        )
                )
        );
        testInvertTree(root);

        // Test 2
        root = buildTree(
                asList(2,
                        List.of(1),
                        List.of(3)
                )
        );
        testInvertTree(root);

        // Test 3
        root = buildTree(
                asList(1,
                        asList(2,
                                List.of(3),
                                List.of(4)
                        ),
                        asList(5,
                                List.of(6),
                                List.of(7)
                        )
                )
        );
        testInvertTree(root);

        root = buildTree(
                asList(1,
                        asList(2,
                                List.of(3),
                                asList(4,
                                        asList(5,
                                                asList(6,
                                                        asList(7,
                                                                asList(8,
                                                                        asList(9,
                                                                                asList(10,
                                                                                        asList(11,
                                                                                                asList(12,
                                                                                                        asList(13,
                                                                                                                asList(14,
                                                                                                                        List.of(15)
                                                                                                                )
                                                                                                        )
                                                                                                )
                                                                                        )
                                                                                )
                                                                        )
                                                                )
                                                        )
                                                )
                                        )
                                )
                        ),
                        asList(5,
                                List.of(6),
                                asList(7,
                                        List.of(8),
                                        asList(9,
                                                List.of(10),
                                                asList(11,
                                                        List.of(12),
                                                        asList(13,
                                                                List.of(14),
                                                                List.of(15)
                                                        )
                                                )
                                        )
                                )
                        )
                )
        );

        testInvertTree(root);
    }

    public static TreeNode invertTree(TreeNode root) {
        if (root == null) return null;

        TreeNode tempLeft = root.left;
        TreeNode tempRight = root.right;

        root.left = invertTree(tempRight);
        root.right = invertTree(tempLeft);

        return root;
    }

    public static void testInvertTree(TreeNode root) {
        System.out.println("-------------------------------------");
        System.out.println("árvore original:");
        System.out.println(root);

        long start = System.nanoTime();
        TreeNode result = invertTree(root);
        long end = System.nanoTime();

        System.out.println("\nárvore invertida:");
        System.out.println(result);
        System.out.println("Execution time: " + (end - start) + "ns");
    }

    @SuppressWarnings("unchecked")
    public static TreeNode buildTree(List<Object> nodes) {
        if (nodes == null || nodes.isEmpty()) return null;

        TreeNode root = new TreeNode((Integer) nodes.get(0));

        if (nodes.size() > 1)
            root.left = buildTree((List<Object>) nodes.get(1));

        if (nodes.size() > 2)
            root.right = buildTree((List<Object>) nodes.get(2));

        return root;
    }
}

File: Ruby/10_binary_tree/226_Invert_Binary_Tree.rb
class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end

  def clone
    TreeNode.new(@val, @left&.clone, @right&.clone)
  end

  def ==(other)
    return true if self.equal?(other)
    return false if other.nil?
    return false unless other.is_a?(TreeNode)

    @val == other.val &&
      @left == other.left &&
      @right == other.right
  end

  def hash
    prime = 31
    hash = 1

    hash *= prime + @val
    hash *= prime + (@left.nil? ? 0 : @left.hash)
    hash *= prime + (@right.nil? ? 0 : @right.hash)

    hash
  end

  def to_s(level = 0, indent = "    ")
    s = "#{indent * level}|-- #{@val}\n"
    s += @left.to_s(level + 1, indent) unless @left.nil?
    s += @right.to_s(level + 1, indent) unless @right.nil?
    s
  end
end

def invert_tree(root)
  return root if root.nil?

  left = invert_tree(root.left)
  right = invert_tree(root.right)

  root.left = right
  root.right = left

  root
end

def build_tree(nodes)
  return nil if nodes.nil? || nodes.empty?
  root = TreeNode.new(nodes[0])
  root.left = build_tree(nodes[1]) if nodes.length > 1
  root.right = build_tree(nodes[2]) if nodes.length > 2
  root
end

def test_invert_tree(root)
  puts "-------------------------------------------------"
  puts "Árvore:"
  puts root.to_s

  start = Time.now
  result = invert_tree(root)
  finish = Time.now

  puts "Árvore invertida:"
  puts result.to_s
  puts "Tempo de execução: #{((finish - start) * 1_000_000_000).to_i}ns"
end


test_invert_tree(build_tree([4, [2, [1], [3]], [7, [6], [9]]]))
test_invert_tree(build_tree([2, [1], [3]]))
test_invert_tree(build_tree([1, [2], [3]]))



